### PR TITLE
Add a flushBuffer API

### PR DESCRIPTION
**Pull request by ColinSullivan1**
This PR adds a flush buffer API to flush the underlying socket (through the socket output stream).

I did use a synchronized API on the connection writer because research seemed to indicate the best performance with a smaller number of threads - at least in older versions of java.  It also fit well with the current architecture since there is no connection level locking.  Instead publish synchronization resolves around the message writer and message queue.  There's a small chance when publishing and immediately doing a flushBuffer, the flushBuffer will beat the message writer, but without major surgery I'm not sure how to fix that.

Regardless, I was able to exceed capacity of the underlying message queue unless I flushed in the test itself so performance doesn't seem to be impacted much.

CC /@scottf @RichardHightower 

Signed-off-by: Colin Sullivan <colin@synadia.com>

### CODE REVIEW

**Review by matthiashanel — COMMENTED** [review]
I'd raise the underlying exception.

**Review by wallyqs — APPROVED** [review]
LGTM

File: src/main/java/io/nats/client/impl/SocketDataPort.java [hunk 1]
```diff
@@ -29,6 +29,9 @@
 
 import io.nats.client.Options;
 
+/**
+ * This class is not theadsafe.  Caller must ensure thread safety.
+ */
 public class SocketDataPort implements DataPort {
```
[the remark this answers is not in the document]
Comment (ColinSullivan1) [reply src/main/java/io/nats/client/impl/SocketDataPort.java:35]: I'm not sure why it was public, perhaps for testing.  Good point though.  I might take a pass through the classes outside of this PR.

File: src/main/java/io/nats/client/Connection.java [hunk 2]
```diff
@@ -382,6 +382,11 @@
      */
     public String createInbox();
 
+    /**
+     * Immediately flushes the underlying connection buffer if the connection is valid.
+     */
+    public void flushBuffer();
```
[the remark this answers is not in the document]
Comment (ColinSullivan1) [reply src/main/java/io/nats/client/Connection.java:388]: I was on the fence, but yeah - probably good to let the user handle that.

**Review by sasbury — APPROVED** [review]
looks ok to me, haven't been in Java for a bit, my only concern is that it works with all the NIO work someone was doing but if this is rebased on that 👍

**Review by matthiashanel — COMMENTED** [review]
LGTM

**Pull request merged**